### ISSUE TITLE
Support DictionaryArrays in Arrow Flight

### ISSUE THREAD

**Issue by alamb** [apache/arrow-rs#3389]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
We are building our application on top of FlightSQL as the transport protocol, with the eventual goal of FlightSQL: #3301 

We make heavy use of DictionaryArrays to encode low cardinality string data. At the moment, we need to decode DictionaryArrays prior  to sending them via Arrow Flight -- see https://github.com/influxdata/influxdb_iox/issues/1318 

**Describe the solution you'd like**
I would like the Flight Client / stuff in arrow-rs to handle Dictionary encoded data efficiently and correctly.

**Describe alternatives you've considered**

**Additional context**
Possibly related / duplicate: https://github.com/apache/arrow-rs/issues/1206

cc @tustvold  and @avantgardnerio

**Comment by alamb** [comment]
Here is my high level plan.

Reference:  https://arrow.apache.org/docs/format/Columnar.html#serialization-and-interprocess-communication-ipc

1. Add an option to `FlightIpcEncoder to "Assign dictionary ids based on pointers", defaults to true
2. If this option is active, the encoder will keep a mapping from actual pointer of the dictionary  to ids
3. The encoder will check any new new dictionary arrays encountered for an existing entry in the map
4. If the entry already exists, the dictionary batch that is transmitted will use the pre-existing entry in the map
5. If the entry does not exist, the dictionary will be transmitted and the entry sent with the batch


There is also the usecase where there are arrays that have the same logical dictionary but the contents are in different actual arrays. While I thought about adding a feature to directly "normalize" these dictionaries by comparing their values, I would like to avoid this for the first version if possible because:

1. It will be non trivially expensive
2. It can be done as a pre-pass over the record batches for systems that want to trade off additional CPU time for less network bandwidtha
3. I can imagine users might want to do other normalizing operations (like combine / coalesce dictionaries) prior to transmission, but that would be more system specific

**Comment by tustvold** [comment]
> Assign dictionary ids based on pointers

I think you will need to keep an owned ArrayData to avoid issues due to pointer reuse, but otherwise this sounds like a good idea. Might want to think about only keeping the last dictionary for each column, to constrain memory growth and keep things simple (you wouldn't need a map)

**Comment by alamb** [comment]
> You might want to think about only keeping the last dictionary for each column, to constrain memory growth and keep things simple (you wouldn't need a map).

Good idea. In fact this may be what the IPC format requires (the dictionary_id is *per field* rather than per *batch* as I was thinking). I believe the spec calls a new dictionary for the same field a "delta dictionary batch"

However, it appears that the arrow ipc reader doesn't actually support this yet

https://github.com/apache/arrow-rs/blob/b371f41f338737b4e214d74b48e18939f5643a84/arrow-ipc/src/reader.rs#L670-L673

I will probably start my work with some arrow-ipc test cleanup (which will convince me we have coverage in the integration tests) and then move on to actually adding support for delta dictionary batches in the ipc reader/writer and then I can add support to flight

**Comment by tustvold** [comment]
> In fact this may be what the IPC format requires

Quite possibly, I think the field is per column and if you want to change dictionary part way through, you need to reserialize it, even if it is actually reverting to a dictionary sent before.

> I believe the spec calls a new dictionary for the same field a "delta dictionary batch"

The [spec](https://arrow.apache.org/docs/format/Columnar.html#dictionary-messages) states

> Alternatively, if isDelta is set to false, then the dictionary replaces the existing dictionary for the same ID.

So I don't think we need to support delta dictionaries in order to unblock this ticket, we can just send the entire dictionary again

**Comment by e-dard** [comment]
I ran into this today when I was surprised to find all my dictionaries were `Utf8` on the server side of the Flight connection. +💯 on getting dictionary transmission support in Flight.

**Comment by alamb** [comment]
We would, of course, welcome contributions in this area. I hope to eventually implement this as it will improve the iox query performance / latency with data from the ingester, but haven't had a chance

**Comment by tustvold** [comment]
I think an option to send the dictionaries every batch sounds reasonable to me, whilst perhaps not "optimal", there are many cases in which case it will be better than the current behaviour :+1:

**Comment by alamb** [comment]
>  I think an option to send the dictionaries every batch sounds reasonable to me, whilst perhaps not "optimal", there are many cases in which case it will be better than the current behaviour, and we can always revisit this later 👍

I agree. Excited to see how that looks @alexwilcoxson-rel . Thank you 🙏

**Comment by alexwilcoxson-rel** [comment]
Opened issue #4895 and PR #4896

**Comment by Jefffrey** [comment]
i think this might be done? see

- https://github.com/apache/arrow-rs/pull/5488
- https://github.com/apache/arrow-rs/pull/4896
